File: scripts/build_fashion_assets.dart
// ignore_for_file: avoid_print

import 'dart:io';

import 'package:image/image.dart' as img;

/// Generates luxury catalog PNGs under assets/fashion/.
void main() {
  final root = Directory('assets/fashion');
  _write(root, 'tops/blazer_beige.png', _product(0xD4, 0xC4, 0xA8, tall: true));
  _write(root, 'tops/corset_cream.png', _product(0xF5, 0xE6, 0xD3, tall: true));
  _write(root, 'tops/satin_top_black.png', _product(0x1A, 0x1A, 0x1A, tall: true));
  _write(root, 'bottoms/skirt_silk_beige.png', _product(0xE8, 0xDC, 0xC8, wide: true));
  _write(root, 'bottoms/trousers_navy.png', _product(0x1E, 0x2A, 0x4A, wide: true));
  _write(root, 'bottoms/skirt_classic_black.png', _product(0x14, 0x14, 0x14, wide: true));
  _write(root, 'outerwear/cape_elegant_ivory.png', _product(0xF7, 0xF2, 0xEA, cape: true));
  _write(root, 'outerwear/cardigan_soft_gray.png', _product(0xC8, 0xC4, 0xBE, tall: true));
  _write(root, 'outerwear/shawl_wedding.png', _product(0xED, 0xE4, 0xF0, wide: true));
  _write(root, 'bags/clutch_silver.png', _product(0xC0, 0xC0, 0xC8, bag: true));
  _write(root, 'bags/bag_black_leather.png', _product(0x1C, 0x1C, 0x1C, bag: true));
  _write(root, 'bags/tote_casual_beige.png', _product(0xD9, 0xCD, 0xB8, bag: true));
  _write(root, 'heels/heels_nude.png', _product(0xE3, 0xC4, 0xA8, shoes: true));
  _write(root, 'heels/heels_black.png', _product(0x11, 0x11, 0x11, shoes: true));
  _write(root, 'jewelry/pearl_earrings.png', _product(0xF5, 0xF0, 0xE8, jewel: true));
  _write(root, 'jewelry/silver_necklace.png', _product(0xB8, 0xB8, 0xC0, jewel: true));
  _write(root, 'jewelry/diamond_bracelet.png', _product(0xE8, 0xE8, 0xF0, jewel: true));
  _write(root, 'jewelry/watch_gold.png', _product(0xC9, 0xA9, 0x62, jewel: true));
  _write(root, 'jewelry/sunglasses_tortoise.png', _product(0x6B, 0x4A, 0x32, jewel: true));
  _write(root, 'scarves/silk_scarf_beige.png', _product(0xE5, 0xD5, 0xBC, wide: true));
  print('Generated fashion assets in ${root.path}');
}

void _write(Directory root, String rel, img.Image image) {
  final file = File('${root.path}/$rel');
  file.parent.createSync(recursive: true);
  file.writeAsBytesSync(img.encodePng(image));
}

img.Image _product(
  int r,
  int g,
  int b, {
  bool tall = false,
  bool wide = false,
  bool cape = false,
  bool bag = false,
  bool shoes = false,
  bool jewel = false,
}) {
  final c = img.Image(width: 280, height: 360);
  for (var y = 0; y < c.height; y++) {
    final t = y / c.height;
    final bgR = (252 + (245 - 252) * t).round();
    final bgG = (252 + (246 - 252) * t).round();
    final bgB = (255 + (250 - 255) * t).round();
    img.fillRect(
      c,
      x1: 0,
      y1: y,
      x2: c.width,
      y2: y + 1,
      color: img.ColorRgb8(bgR, bgG, bgB),
    );
  }

  img.fillCircle(
    c,
    x: 140,
    y: 280,
    radius: 62,
    color: img.ColorRgba8(0, 0, 0, 24),
  );

  final color = img.ColorRgb8(r, g, b);
  final shade = img.ColorRgb8(
    (r * 0.82).round().clamp(0, 255),
    (g * 0.82).round().clamp(0, 255),
    (b * 0.82).round().clamp(0, 255),
  );

  if (shoes) {
    img.fillRect(c, x1: 70, y1: 210, x2: 130, y2: 255, radius: 10, color: color);
    img.fillRect(c, x1: 150, y1: 215, x2: 210, y2: 260, radius: 10, color: shade);
  } else if (bag) {
    img.fillRect(c, x1: 88, y1: 150, x2: 192, y2: 250, radius: 18, color: color);
    img.fillRect(c, x1: 118, y1: 125, x2: 162, y2: 150, radius: 8, color: shade);
  } else if (jewel) {
    img.fillCircle(c, x: 140, y: 190, radius: 36, color: color);
    img.fillCircle(c, x: 128, y: 178, radius: 10, color: img.ColorRgb8(255, 255, 255));
  } else if (cape) {
    img.fillRect(c, x1: 52, y1: 95, x2: 228, y2: 290, radius: 28, color: color);
    img.fillRect(c, x1: 118, y1: 85, x2: 162, y2: 120, radius: 8, color: shade);
  } else if (wide) {
    img.fillRect(c, x1: 72, y1: 140, x2: 208, y2: 285, radius: 22, color: color);
  } else {
    img.fillRect(c, x1: 82, y1: tall ? 95 : 120, x2: 198, y2: tall ? 270 : 285, radius: 20, color: color);
    img.fillRect(c, x1: 128, y1: tall ? 110 : 130, x2: 152, y2: tall ? 250 : 270, radius: 6, color: shade);
  }

  return c;
}

File: scripts/build_fashion_assets.py
#!/usr/bin/env python3
"""Generate luxury fashion catalog PNGs for MIRA recommendation cards."""
from __future__ import annotations

import math
from pathlib import Path

from PIL import Image, ImageDraw, ImageFilter

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "assets" / "fashion"
SIZE = (280, 360)


def _hex(color: str) -> tuple[int, int, int]:
    c = color.lstrip("#")
    return tuple(int(c[i : i + 2], 16) for i in (0, 2, 4))


def _lerp(a: int, b: int, t: float) -> int:
    return int(a + (b - a) * t)


def _shade(base: tuple[int, int, int], delta: float) -> tuple[int, int, int]:
    return tuple(max(0, min(255, int(v + delta * 255))) for v in base)


def _canvas() -> tuple[Image.Image, ImageDraw.ImageDraw]:
    img = Image.new("RGBA", SIZE, (255, 255, 255, 0))
    draw = ImageDraw.Draw(img)
    # soft studio backdrop
    for y in range(SIZE[1]):
        t = y / SIZE[1]
        c = (
            _lerp(252, 245, t),
            _lerp(252, 246, t),
            _lerp(255, 250, t),
            255,
        )
        draw.line([(0, y), (SIZE[0], y)], fill=c)
    return img, draw


def _shadow(draw: ImageDraw.ImageDraw, box: tuple[int, int, int, int]) -> None:
    x0, y0, x1, y1 = box
    cx = (x0 + x1) // 2
    cy = y1 + 8
    draw.ellipse([cx - 70, cy - 8, cx + 70, cy + 18], fill=(0, 0, 0, 28))


def _save(img: Image.Image, rel: str) -> None:
    path = OUT / rel
    path.parent.mkdir(parents=True, exist_ok=True)
    img.convert("RGB").save(path, quality=92)


def draw_blazer(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (70, 90, 210, 280))
    body = [(78, 88), (202, 88), (214, 290), (66, 290)]
    draw.polygon(body, fill=base)
    draw.polygon(body, outline=_shade(base, -0.18))
    draw.polygon([(140, 88), (108, 170), (140, 155)], fill=_shade(base, 0.08))
    draw.polygon([(140, 88), (172, 170), (140, 155)], fill=_shade(base, 0.04))
    draw.ellipse([132, 178, 148, 194], fill=_shade(base, -0.25))
    _save(img, path)


def draw_top(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (80, 110, 200, 250))
    draw.polygon([(95, 105), (185, 105), (198, 250), (82, 250)], fill=base)
    draw.arc([95, 95, 185, 145], 200, 340, fill=_shade(base, -0.15), width=3)
    draw.line([(140, 105), (140, 250)], fill=_shade(base, -0.1), width=2)
    _save(img, path)


def draw_skirt(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (70, 150, 210, 300))
    draw.polygon([(95, 145), (185, 145), (220, 305), (60, 305)], fill=base)
    for i in range(6):
        x = 95 + i * 18
        draw.line([(x, 145), (x + 8, 305)], fill=_shade(base, -0.06), width=2)
    _save(img, path)


def draw_trousers(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (80, 140, 200, 310))
    draw.polygon([(95, 135), (185, 135), (190, 190), (150, 310), (130, 310), (90, 190)], fill=base)
    draw.line([(140, 135), (140, 310)], fill=_shade(base, -0.12), width=2)
    _save(img, path)


def draw_cape(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (50, 100, 230, 290))
    draw.polygon([(140, 85), (55, 120), (35, 300), (245, 300), (225, 120)], fill=base)
    draw.arc([110, 80, 170, 130], 200, 340, fill=_shade(base, -0.12), width=3)
    _save(img, path)


def draw_cardigan(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (70, 95, 210, 285))
    draw.rounded_rectangle([72, 92, 208, 285], radius=18, fill=base)
    for x in (108, 140, 172):
        draw.line([(x, 110), (x, 270)], fill=_shade(base, -0.08), width=3)
    _save(img, path)


def draw_shawl(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (60, 120, 220, 280))
    draw.polygon([(60, 130), (220, 150), (200, 280), (80, 260)], fill=base)
    draw.line([(60, 130), (220, 150)], fill=_shade(base, -0.1), width=4)
    _save(img, path)


def draw_bag(path: str, color: str, clutch: bool = False) -> None:
    img, draw = _canvas()
    base = _hex(color)
    if clutch:
        _shadow(draw, (85, 150, 195, 230))
        draw.rounded_rectangle([88, 155, 192, 225], radius=14, fill=base)
        draw.arc([118, 130, 162, 165], 200, 340, fill=_shade(base, -0.2), width=4)
    else:
        _shadow(draw, (75, 130, 205, 260))
        draw.rounded_rectangle([82, 145, 198, 255], radius=20, fill=base)
        draw.arc([108, 105, 172, 155], 200, 340, fill=_shade(base, -0.25), width=5)
        if "tote" in path:
            draw.rectangle([82, 145, 198, 165], fill=_shade(base, -0.08))
    _save(img, path)


def draw_heels(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (70, 180, 210, 290))
    # left shoe
    draw.polygon([(75, 220), (130, 220), (145, 255), (70, 255)], fill=base)
    draw.polygon([(130, 220), (130, 175), (138, 175), (145, 255)], fill=_shade(base, -0.12))
    # right shoe
    draw.polygon([(150, 225), (205, 225), (220, 260), (145, 260)], fill=base)
    draw.polygon([(205, 225), (205, 180), (213, 180), (220, 260)], fill=_shade(base, -0.12))
    _save(img, path)


def draw_pearls(path: str) -> None:
    img, draw = _canvas()
    _shadow(draw, (90, 150, 190, 210))
    for i, x in enumerate(range(108, 173, 16)):
        y = 168 + (i % 2) * 6
        draw.ellipse([x, y, x + 14, y + 14], fill=(245, 240, 232))
        draw.ellipse([x + 3, y + 2, x + 9, y + 8], fill=(255, 255, 255, 180))
    draw.line([(115, 175), (165, 175)], fill=(200, 190, 170), width=2)
    _save(img, path)


def draw_necklace(path: str) -> None:
    img, draw = _canvas()
    _shadow(draw, (80, 160, 200, 230))
    draw.arc([90, 130, 190, 230], 200, 340, fill=(180, 180, 190), width=4)
    draw.ellipse([132, 198, 148, 214], fill=(210, 210, 220))
    for x in (108, 124, 156, 172):
        draw.ellipse([x, 176, x + 8, x + 8 if False else 184], fill=(190, 190, 200))
    _save(img, path)


def draw_bracelet(path: str) -> None:
    img, draw = _canvas()
    _shadow(draw, (70, 170, 210, 230))
    draw.arc([85, 165, 195, 235], 30, 150, fill=(210, 210, 225), width=10)
    for a in range(0, 120, 15):
        rad = math.radians(30 + a)
        cx = 140 + int(52 * math.cos(rad))
        cy = 200 + int(28 * math.sin(rad))
        draw.ellipse([cx - 4, cy - 4, cx + 4, cy + 4], fill=(255, 255, 255))
    _save(img, path)


def draw_scarf(path: str, color: str) -> None:
    img, draw = _canvas()
    base = _hex(color)
    _shadow(draw, (60, 130, 220, 290))
    draw.polygon([(120, 95), (170, 110), (150, 300), (90, 280)], fill=base)
    draw.polygon([(120, 95), (95, 280), (110, 290), (130, 120)], fill=_shade(base, -0.06))
    _save(img, path)


def draw_watch(path: str) -> None:
    img, draw = _canvas()
    _shadow(draw, (95, 150, 185, 230))
    draw.rounded_rectangle([108, 165, 172, 215], radius=16, fill=(201, 169, 98))
    draw.ellipse([118, 175, 162, 205], fill=(250, 248, 242))
    draw.line([(140, 185), (140, 178)], fill=(40, 40, 40), width=2)
    draw.line([(140, 185), (152, 192)], fill=(40, 40, 40), width=2)
    draw.rectangle([130, 155, 150, 165], fill=(180, 150, 80))
    _save(img, path)


def draw_sunglasses(path: str) -> None:
    img, draw = _canvas()
    _shadow(draw, (70, 170, 210, 220))
    draw.rounded_rectangle([78, 175, 128, 210], radius=12, fill=(60, 40, 28, 220))
    draw.rounded_rectangle([152, 175, 202, 210], radius=12, fill=(60, 40, 28, 220))
    draw.line([(128, 190), (152, 190)], fill=(50, 35, 25), width=4)
    draw.line([(78, 190), (58, 182)], fill=(50, 35, 25), width=3)
    _save(img, path)


def main() -> None:
    draw_blazer("tops/blazer_beige.png", "#D4C4A8")
    draw_top("tops/corset_cream.png", "#F5E6D3")
    draw_top("tops/satin_top_black.png", "#1A1A1A")
    draw_skirt("bottoms/skirt_silk_beige.png", "#E8DCC8")
    draw_trousers("bottoms/trousers_navy.png", "#1E2A4A")
    draw_skirt("bottoms/skirt_classic_black.png", "#141414")
    draw_cape("outerwear/cape_elegant_ivory.png", "#F7F2EA")
    draw_cardigan("outerwear/cardigan_soft_gray.png", "#C8C4BE")
    draw_shawl("outerwear/shawl_wedding.png", "#EDE4F0")
    draw_bag("bags/clutch_silver.png", "#C0C0C8", clutch=True)
    draw_bag("bags/bag_black_leather.png", "#1C1C1C")
    draw_bag("bags/tote_casual_beige.png", "#D9CDB8")
    draw_heels("heels/heels_nude.png", "#E3C4A8")
    draw_heels("heels/heels_black.png", "#111111")
    draw_pearls("jewelry/pearl_earrings.png")
    draw_necklace("jewelry/silver_necklace.png")
    draw_bracelet("jewelry/diamond_bracelet.png")
    draw_scarf("scarves/silk_scarf_beige.png", "#E5D5BC")
    draw_watch("jewelry/watch_gold.png")
    draw_sunglasses("jewelry/sunglasses_tortoise.png")
    print(f"Generated fashion assets under {OUT}")


if __name__ == "__main__":
    main()
